sdk/monitor/monitor-opentelemetry-exporter/src/utils/spanUtils.ts

```diff
@@ -1,75 +1,87 @@
+    tags[KnownContextTagKeys.AiOperationParentId] = span.parentSpanId;
   }
   if (span.resource && span.resource.attributes) {
-    const serviceName = span.resource.attributes[ResourceAttributes.SERVICE_NAME];
-    const serviceNamespace = span.resource.attributes[ResourceAttributes.SERVICE_NAMESPACE];
-    const serviceInstanceId = span.resource.attributes[ResourceAttributes.SERVICE_INSTANCE_ID];
+    const serviceName = span.resource.attributes[SemanticResourceAttributes.SERVICE_NAME];
+    const serviceNamespace = span.resource.attributes[SemanticResourceAttributes.SERVICE_NAMESPACE];
     if (serviceName) {
       if (serviceNamespace) {
-        tags[AI_CLOUD_ROLE] = `${serviceNamespace}.${serviceName}`;
+        tags[KnownContextTagKeys.AiCloudRole] = `${serviceNamespace}.${serviceName}`;
       } else {
-        tags[AI_CLOUD_ROLE] = String(serviceName);
+        tags[KnownContextTagKeys.AiCloudRole] = String(serviceName);
       }
     }
+    const serviceInstanceId =
+      span.resource.attributes[SemanticResourceAttributes.SERVICE_INSTANCE_ID];
     if (serviceInstanceId) {
-      tags[AI_CLOUD_ROLE_INSTACE] = String(serviceInstanceId);
+      tags[KnownContextTagKeys.AiCloudRoleInstance] = String(serviceInstanceId);
+    } else {
+      tags[KnownContextTagKeys.AiCloudRoleInstance] = os && os.hostname();
+    }
+    const endUserId = span.resource.attributes[SemanticAttributes.ENDUSER_ID];
+    if (endUserId) {
+      tags[KnownContextTagKeys.AiUserId] = String(endUserId);
     }
   }
-
-  // @todo: is this for RequestData only?
-  if (
-    (span.kind === SpanKind.SERVER || span.kind === SpanKind.CONSUMER) &&
-    span.attributes[GRPC_METHOD]
-  ) {
-    tags[AI_OPERATION_NAME] = String(span.attributes[GRPC_METHOD]);
-  }
-  if (
-    (span.kind === SpanKind.SERVER || span.kind === SpanKind.CONSUMER) &&
-    span.attributes[HTTP_METHOD] &&
-    span.attributes[HTTP_ROUTE]
-  ) {
-    tags[AI_OPERATION_NAME] = `${span.attributes[HTTP_METHOD] as string} ${span.attributes[
-      HTTP_ROUTE
-    ] as string}`;
+  // HTTP Spans
+  const httpMethod = span.attributes[SemanticAttributes.HTTP_METHOD];
+  if (httpMethod) {
+    let httpClientIp = null;
+    let netPeerIp = null;
+    if (span.resource && span.resource.attributes) {
+      httpClientIp = span.resource.attributes[SemanticAttributes.HTTP_CLIENT_IP];
+      netPeerIp = span.resource.attributes[SemanticAttributes.NET_PEER_IP];
+    }
+    if (span.kind === SpanKind.SERVER) {
+      tags[KnownContextTagKeys.AiOperationName] = `${httpMethod as string} ${span.name as string}`;
+      if (httpClientIp) {
+        tags[KnownContextTagKeys.AiLocationIp] = String(httpClientIp);
+      } else if (netPeerIp) {
+        tags[KnownContextTagKeys.AiLocationIp] = String(netPeerIp);
+      }
+    } else {
+      tags[KnownContextTagKeys.AiOperationName] = span.name;
+      if (netPeerIp) {
+        tags[KnownContextTagKeys.AiLocationIp] = String(netPeerIp);
```

Comment:
There isn't a definition for `ai.location.ip` for non-server spans is there?

Reply:
ai.location.ip Mapping:
In case of http server spans: httpClientIp (if available). If httpClientIp is not available, then netPeerIp (if available).
For non-http server spans: netPeerIp (if available).

https://github.com/microsoft/common-schema/blob/main/Mappings/AzureMonitor-AI.md#mapping-rule

Reply:
Right, I'm referring to the case where it is not a server span. Spec only covers http/non-http SERVER spans.

Reply:
Updated this